### PR TITLE
Implemented sliding window to calculate values changes/s

### PR DESCRIPTION
**Pull request by koepalex**
* Extended IMessageTrigger 
  *  property `DataChangesCountLastMinute` that calculates all OPC UA data changes in the last 60 seconds
  *  property `ValueChangesCountLastMinute` that calculates all OPC UA value changes in the last 60 seconds
* Updated diagnostic log with values of last 60 seconds
* Reset diagnostic log after first data is received, to align timings

image

[ADO #10205715](https://msazure.visualstudio.com/One/_workitems/edit/10205715/)
Improved logging for #912

### CODE REVIEW

File: components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Edge.Publisher/src/Engine/WriterGroupMessageSource.cs [hunk 1]
```diff
@@ -30,11 +30,56 @@ public class WriterGroupMessageTrigger : IMessageTrigger, IDisposable {
...
+            private set
+            {
+                var difference = Math.Abs(value);
+                var now = DateTime.UtcNow;
+                var oneMinuteTimespan = TimeSpan.FromMinutes(1);
+                _dataChanges.Add(now, difference);
+
+                var dataChangesToDelete = _dataChanges.Where(kvp => now - kvp.Key > oneMinuteTimespan).Select(kvp => kvp.Key);
+                foreach (var dataChangeToDelete in dataChangesToDelete) {
+                    _dataChanges.Remove(dataChangeToDelete);
```
Comment (mregen) [inline components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Edge.Publisher/src/Engine/WriterGroupMessageSource.cs]: >_dataChanges.Remove(dataChangeToDelete) [](http://example.com/codeflow?start=20&length=39)

does it need a lock? can it collide with a get call?

File: components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Edge.Publisher/src/Engine/WriterGroupMessageSource.cs [hunk 2]
```diff
@@ -30,11 +30,56 @@ public class WriterGroupMessageTrigger : IMessageTrigger, IDisposable {
         public int NumberOfConnectionRetries => _subscriptions?.FirstOrDefault()?.
             Subscription?.NumberOfConnectionRetries ?? 0;
 
+        private IDictionary<DateTime, int> _valueChanges = new Dictionary<DateTime, int>(100);
```
Comment (mregen) [inline components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Edge.Publisher/src/Engine/WriterGroupMessageSource.cs]: >new Dictionary<DateTime, int>(100) [](http://example.com/codeflow?start=59&length=34)

idea: use a histogram array (ringbuffer which counts e.g. 10 s blocks). Then no need to manage dictionary content

File: components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Edge.Publisher/src/Engine/DataFlowProcessingEngine.cs [hunk 3]
```diff
@@ -222,11 +224,11 @@ public class DataFlowProcessingEngine : IProcessingEngine, IDisposable {
             kDataChangesCount.WithLabels(deviceId, moduleId, Name)
                 .Set(_messageTrigger.DataChangesCount);
             kDataChangesPerSecond.WithLabels(deviceId, moduleId, Name)
-                .Set(dataChangesPerSec);
+                .Set(dataChangesPerSecLastMin);
             kValueChangesCount.WithLabels(deviceId, moduleId, Name)
                 .Set(_messageTrigger.ValueChangesCount);
             kValueChangesPerSecond.WithLabels(deviceId, moduleId, Name)
-                .Set(valueChangesPerSec);
+                .Set(valueChangesPerSecLastMin);
```
Comment (karok2m) [inline components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Edge.Publisher/src/Engine/DataFlowProcessingEngine.cs]: I would suggest to keep existing Prometheus metrics as they are and introduce new ones for the sliding intervals. #Closed

Comment (koepalex) [reply components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Edge.Publisher/src/Engine/DataFlowProcessingEngine.cs]: changed

File: components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Edge.Publisher/src/Engine/DataFlowProcessingEngine.cs [hunk 4]
```diff
@@ -175,9 +177,9 @@ public class DataFlowProcessingEngine : IProcessingEngine, IDisposable {
             var diagInfo = new StringBuilder();
             diagInfo.AppendLine("\n  DIAGNOSTICS INFORMATION for          : {host}");
             diagInfo.AppendLine("  # Ingestion duration                 : {duration,14:dd\\:hh\\:mm\\:ss} (dd:hh:mm:ss)");
-            string dataChangesPerSecFormatted = _messageTrigger.DataChangesCount > 0 && totalDuration > 0 ? $"({dataChangesPerSec:0.##}/s)" : "";
+            string dataChangesPerSecFormatted = _messageTrigger.DataChangesCount > 0 && totalDuration > 0 ? $"(All time ~{dataChangesPerSec:0.##}/s; {_messageTrigger.DataChangesCountLastMinute.ToString("D2")} in last 60s ~{dataChangesPerSecLastMin:0.##}/s)" : "";
```
Comment (karok2m) [inline components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Edge.Publisher/src/Engine/DataFlowProcessingEngine.cs]: @koepalex, can you please break lines `180` and `182` into several lines ? They seem to longer than `120` characters.

File: components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Edge.Publisher/src/Engine/WriterGroupMessageSource.cs [hunk 5]
```diff
@@ -30,11 +30,56 @@ public class WriterGroupMessageTrigger : IMessageTrigger, IDisposable {
...
+                var valuesToDelete = _valueChanges.Where(kvp => now - kvp.Key > oneMinuteTimespan).Select(kvp => kvp.Key);
+                foreach (var valueToDelete in valuesToDelete) {
+                    _valueChanges.Remove(valueToDelete);
+                }
+            }
+        }
+
+        private int _valueChangesCount = 0;
+        /// <inheritdoc/>
+        public int ValueChangesCount { get { return _valueChangesCount; } private set { var difference = (value - _valueChangesCount);  _valueChangesCount = value; ValueChangesCountLastMinute = difference; } }
```
Comment (karok2m) [inline components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Edge.Publisher/src/Engine/WriterGroupMessageSource.cs]: @koepalex, please break lines `57` and `82` as well into several lines if they are longer than `120` characters. It is kind of hard to tell on GitHub if they are or not.

File: components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Edge.Publisher/src/Engine/WriterGroupMessageSource.cs [hunk 6]
```diff
@@ -30,11 +30,56 @@ public class WriterGroupMessageTrigger : IMessageTrigger, IDisposable {
...
+            get
+            {
+                return _dataChanges.Values.Sum();
+            }
+            private set
+            {
+                var difference = Math.Abs(value);
+                var now = DateTime.UtcNow;
+                var oneMinuteTimespan = TimeSpan.FromMinutes(1);
+                _dataChanges.Add(now, difference);
```
Comment (karok2m) [inline components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Edge.Publisher/src/Engine/WriterGroupMessageSource.cs]: Do we need to worry about the case when there is already a value for `now` entry ? I guess not, but I wanted to raise the question. #Closed

Comment (karok2m) [reply components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Edge.Publisher/src/Engine/WriterGroupMessageSource.cs]: The same question applies for `_valueChanges` on line `46`.

File: components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Edge.Publisher/src/Engine/WriterGroupMessageSource.cs (hunk 2 above)
Comment (koepalex) [reply components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Edge.Publisher/src/Engine/WriterGroupMessageSource.cs]: good idea, changed implementation to ring buffer

File: components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Edge.Publisher/src/Engine/DataFlowProcessingEngine.cs (hunk 4 above)
Comment (koepalex) [reply components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Edge.Publisher/src/Engine/DataFlowProcessingEngine.cs]: added breaks

File: components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Edge.Publisher/src/Engine/WriterGroupMessageSource.cs (hunk 5 above)
Comment (koepalex) [reply components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Edge.Publisher/src/Engine/WriterGroupMessageSource.cs]: added line breaks

File: components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Edge.Publisher/src/Engine/WriterGroupMessageSource.cs (hunk 6 above)
Comment (koepalex) [reply components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Edge.Publisher/src/Engine/WriterGroupMessageSource.cs]: changed implementation to ring buffer, not valid anymore

**Comment by koepalex** [comment]
The ring buffer implementation is really fast:
image

File: components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Edge.Publisher/src/Engine/WriterGroupMessageSource.cs (hunk 1 above)
Comment (koepalex) [reply components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Edge.Publisher/src/Engine/WriterGroupMessageSource.cs]: outdated

File: components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Edge.Publisher/src/Engine/WriterGroupMessageSource.cs [hunk 7]
```diff
@@ -31,20 +31,70 @@ public class WriterGroupMessageTrigger : IMessageTrigger, IDisposable {
...
+        /// <summary>
+        /// Helper function to distribute values over array based on time
+        /// </summary>
+        private void IncreaseRingBuffer(ulong[] array, ref int lastPointer, int bucketWidth, ulong difference) {
+            var indexPointer = DateTime.UtcNow.Second % bucketWidth;
+            while (lastPointer != indexPointer && (lastPointer + 1) < array.Length) {
+                lastPointer++;
+                array[lastPointer] = 0;
+            }
+
```
Comment (mregen) [inline components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Edge.Publisher/src/Engine/WriterGroupMessageSource.cs:112]: the wrap around case seems not quite right:

while (lastPointer != indexPointer)
 {
                lastPointer = (lastPointer+1) % bucketWidth;
                array[lastPointer] = 0;
            }

File: components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Edge.Publisher/src/Engine/WriterGroupMessageSource.cs [hunk 8]
```diff
@@ -31,20 +31,70 @@ public class WriterGroupMessageTrigger : IMessageTrigger, IDisposable {
...
+            for(int index = 0; index< array.Length; index++) {
+                sum += array[index];
+            }
+            return sum;
+        }
+
+        /// <summary>
+        /// Helper function to distribute values over array based on time
+        /// </summary>
+        private void IncreaseRingBuffer(ulong[] array, ref int lastPointer, int bucketWidth, ulong difference) {
```
Comment (mregen) [inline components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Edge.Publisher/src/Engine/WriterGroupMessageSource.cs]: >IncreaseRingBuffer [](http://example.com/codeflow?start=21&length=18)

good stuff :sunglasses:  #Closed

File: components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Edge.Publisher/src/Engine/WriterGroupMessageSource.cs [hunk 9]
```diff
@@ -31,20 +31,70 @@ public class WriterGroupMessageTrigger : IMessageTrigger, IDisposable {
...
+                var difference = value - _dataChangesCount;
+                _dataChangesCount = value;
+                DataChangesCountLastMinute = difference;
+            }
+        }
+
+        /// <summary>
+        /// Iterates the array and add up all values
+        /// </summary>
+        private static ulong CalculateSumForRingBuffer(ulong[] array) {
```
Comment (mregen) [inline components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Edge.Publisher/src/Engine/WriterGroupMessageSource.cs]: >CalculateSumForRingBuffer [](http://example.com/codeflow?start=29&length=25)

sliding window needs to be updated here too...if increase buffer is not called for a few seconds, stale slots must be erased #Closed

File: components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Edge.Publisher/src/Engine/WriterGroupMessageSource.cs [hunk 10]
```diff
@@ -31,20 +31,70 @@ public class WriterGroupMessageTrigger : IMessageTrigger, IDisposable {
...
+            }
+            return sum;
+        }
+
+        /// <summary>
+        /// Helper function to distribute values over array based on time
+        /// </summary>
+        private void IncreaseRingBuffer(ulong[] array, ref int lastPointer, int bucketWidth, ulong difference) {
+            var indexPointer = DateTime.UtcNow.Second % bucketWidth;
+            while (lastPointer != indexPointer && (lastPointer + 1) < array.Length) {
```
Comment (mregen) [inline components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Edge.Publisher/src/Engine/WriterGroupMessageSource.cs]: >while (lastPointer ! [](http://example.com/codeflow?start=12&length=20)

one more special case. if last update was > bucketwidth seconds in the past, erase the array completely #Closed

**Review by mregen — CHANGES_REQUESTED** [review]
:clock1:

File: components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Edge.Publisher/src/Engine/WriterGroupMessageSource.cs [hunk 11]
```diff
@@ -31,20 +31,97 @@ public class WriterGroupMessageTrigger : IMessageTrigger, IDisposable {
...
+        /// Empty the ring buffer buckets if necessary
+        /// </summary>
+        private static void CleanStaleRingBufferBuckets(ulong[] array, ref int lastPointer, int bucketWidth, int indexPointer, DateTime lastWrite, DateTime now) {
+
+            // if last update was > bucketsize seconds in the past delete whole array
+            if (lastWrite != DateTime.MinValue) {
+                var deleteWholeArray = (now - lastWrite).TotalSeconds > bucketWidth;
+                if (deleteWholeArray) {
+                    Array.Clear(array, 0, array.Length);
+                    return;
```
Comment (karok2m) [inline components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Edge.Publisher/src/Engine/WriterGroupMessageSource.cs]: We need to update `lastPointer` in this case as well. I believe that it should be assigned the value of `indexPointer`. #Closed

File: components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Edge.Publisher/src/Engine/WriterGroupMessageSource.cs (hunk 10 above)
Comment (koepalex) [reply components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Edge.Publisher/src/Engine/WriterGroupMessageSource.cs]: implemented

File: components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Edge.Publisher/src/Engine/WriterGroupMessageSource.cs (hunk 9 above)
Comment (koepalex) [reply components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Edge.Publisher/src/Engine/WriterGroupMessageSource.cs]: done

File: components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Edge.Publisher/src/Engine/WriterGroupMessageSource.cs (hunk 8 above)
Comment (koepalex) [reply components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Edge.Publisher/src/Engine/WriterGroupMessageSource.cs]: yes I think so :)

File: components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Edge.Publisher/src/Engine/WriterGroupMessageSource.cs (hunk 7 above)
Comment (koepalex) [reply components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Edge.Publisher/src/Engine/WriterGroupMessageSource.cs:112]: changed

File: components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Edge.Publisher/src/Engine/WriterGroupMessageSource.cs (hunk 11 above)
Comment (koepalex) [reply components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Edge.Publisher/src/Engine/WriterGroupMessageSource.cs]: no it is updated in line 94, when the write happens

File: components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Edge.Publisher/src/Engine/WriterGroupMessageSource.cs [hunk 12]
```diff
@@ -31,20 +31,97 @@ public class WriterGroupMessageTrigger : IMessageTrigger, IDisposable {
...
+        /// <summary>
+        /// Helper function to distribute values over array based on time
+        /// </summary>
+        private static void IncreaseRingBuffer(ulong[] array, ref int lastPointer, int bucketWidth, ulong difference, ref DateTime lastWriteTime) {
+            var now = DateTime.UtcNow;
+            var indexPointer = now.Second % bucketWidth;
+            CleanStaleRingBufferBuckets(array, ref lastPointer, bucketWidth, indexPointer, lastWriteTime, now);
+
+            array[indexPointer] += difference;
+            lastPointer = indexPointer;
```
Comment (mregen) [inline components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Edge.Publisher/src/Engine/WriterGroupMessageSource.cs]: >lastPointer = indexPointer; [](http://example.com/codeflow?start=12&length=27)

this assignment is obsolete, instead as discussed

File: components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Edge.Publisher/src/Engine/WriterGroupMessageSource.cs (hunk 11 above)
Comment (mregen) [reply components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Edge.Publisher/src/Engine/WriterGroupMessageSource.cs]: changed :ok:

Comment (koepalex) [reply components/opc-ua/src/Microsoft.Azure.IIoT.OpcUa.Edge.Publisher/src/Engine/WriterGroupMessageSource.cs]: anyway changed for better readability

**Review by mregen — APPROVED** [review]
:shipit:

**Pull request merged**